### PR TITLE
 Drop invisible SemanticsNodes from tree

### PR DESCRIPTION
**Pull request by goderbauer**
A node is invisible if it is outside of the bounds of the screen and if it is not merged into its (partially) visible parent.

Also in this PR: only set `wasAffectedByClip` to true if the nodes has actually been clipped.

This change is required for https://github.com/flutter/flutter/issues/12275.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 1]
```diff
@@ -285,9 +285,20 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
     }
   }
 
-  /// Whether [rect] might have been influenced by clips applied by ancestors.
+  /// Whether [rect] was clipped by ancestors.
   bool wasAffectedByClip = false;
 
+  /// Whether the node is invisible.
+  ///
+  /// A node, whose [rect] is outside of the bounds of the screen and hence not
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: extraneous comma

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 2]
```diff
@@ -285,9 +285,20 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+  /// reachable for users is considered invisible if its semantic information
+  /// are not merged into a (partially) visible parent as indicated by
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: is not

File: packages/flutter/test/material/scaffold_test.dart [hunk 3]
```diff
@@ -501,11 +501,11 @@ void main() {
...
-      body: new Semantics(label: bodyLabel, child: new Container()),
-      persistentFooterButtons: <Widget>[new Semantics(label: persistentFooterButtonLabel, child: new Container())],
-      bottomNavigationBar: new Semantics(label: bottomNavigationBarLabel, child: new Container()),
-      floatingActionButton: new Semantics(label: floatingActionButtonLabel, child: new Container()),
-      drawer: new Drawer(child:new Semantics(label: drawerLabel, child: new Container())),
+      body: const Text(bodyLabel),
+      persistentFooterButtons: <Widget>[const Text(persistentFooterButtonLabel)],
+      bottomNavigationBar: const Text(bottomNavigationBarLabel),
+      floatingActionButton: const Text(floatingActionButtonLabel),
+      drawer: const Drawer(child:const Text(drawerLabel)),
```
Comment (Hixie) [inline packages/flutter/test/material/scaffold_test.dart]: this is fine, but why this change?

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -605,10 +605,10 @@ class _SemanticsGeometry {
     semantics.transform = _transform;
     if (_clipRect != null) {
       semantics.rect = _clipRect.intersect(rendering.semanticBounds);
-      semantics.wasAffectedByClip = true;
+      semantics.wasAffectedByClip = semantics.rect != rendering.semanticBounds;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: Should probably locally cache semantics.rect and rendering.semanticBounds.

File: packages/flutter/lib/src/rendering/object.dart [hunk 5]
```diff
@@ -605,10 +605,10 @@ class _SemanticsGeometry {
...
     } else {
       semantics.rect = rendering.semanticBounds;
-      semantics.wasAffectedByClip = parentSemantics?.wasAffectedByClip ?? false;
+      semantics.wasAffectedByClip = false;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: wasAffectedByClip is defined as being "Whether rect might have been influenced by clips applied by ancestors", so why is this wrong?

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/test/material/scaffold_test.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/test/material/scaffold_test.dart]: Previously, the `Semantics` widgets were creating `SemanticsNode`s with a size of `Rect.zero`. With the new definition of visibility introduced in this PRs, those `SemanticsNode`s are invisible and are no longer allowed in the tree (ensured with asserts). Rather than giving the `Container` in the old code a width and a height, I felt like the code would be easier to read if I instead replace the entire thing with a `Text` widget that has an intrinsic width/height.

File: packages/flutter/lib/src/rendering/object.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/object.dart]: I am slightly redefining the meaning of wasAffectedByClip as indicated by the new doc comment and the description of this PR. It's only set to true if the rect of the SemanticsNode has actually changed due to clipping. I find that information more useful. Previously, when there was one clipped node in the semantics tree all of its descendants would be marked as clipped. That made it impossible to figure out which children actually had their size changed due to clipping.

Comment (Hixie) [reply packages/flutter/lib/src/rendering/object.dart]: That's totally fine but then you should update the documentation to match :-)

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/object.dart]: Ohh, I did: https://github.com/flutter/flutter/pull/12358/files#diff-647e4aa016f360586e38d3cdf5bad6f0R288

I will make that even clearer by adding some more information from my previous comment in this thread.
